python/tests/unittests/test_kernel_builder.py

```diff
@@ -978,6 +978,71 @@ def test_fermionic_swap_op():
     assert np.isclose(val3, 0.0, atol=1e-6)
     assert np.isclose(val4, 0.0, atol=1e-6)
 
+def test_draw():
+    print()
+    bar, q = cudaq.make_kernel(cudaq.qreg)
+    bar.rx(np.e, q[0])
+    bar.ry(np.pi, q[1])
+    bar.rz(-np.pi, q[2]) # how to do adj in builder?
+
+    zaz, q = cudaq.make_kernel(cudaq.qubit)
+    zaz.sdg(q)
+
+    kernel = cudaq.make_kernel()
+    q = kernel.qalloc(4)
+
+    kernel.h(q); # Broadcast
+    kernel.cx(q[0], q[1])
+    kernel.cy([q[0], q[1]], q[2])
+    kernel.cy([q[2], q[0]], q[1])
+    kernel.cy([q[1], q[2]], q[0])
+    kernel.z(q[2])
+
+    kernel.r1(3.14159, q[0])
+    kernel.tdg(q[1])
+    kernel.s(q[2])
+
+    kernel.swap(q[0], q[2])
+    kernel.swap(q[1], q[2])
+    kernel.swap(q[0], q[1])
+    kernel.swap(q[0], q[2])
+    kernel.swap(q[1], q[2])
+    kernel.cswap(q[3], q[0], q[1])
+    kernel.cswap([q[0], q[3]], q[1], q[2])
+    kernel.cswap(q[1], q[0], q[3])
+    kernel.cswap([q[1], q[2]], q[0], q[3])
+    kernel.apply_call(bar, q)
+    kernel.control(zaz, q[1], q[0])
+    kernel.adjoint(bar, q)
+
+    circuit = cudaq.draw(kernel)
+    print(circuit)
+    expected_str = '''
+     ╭───╮               ╭───╮╭───────────╮                          ╭───────╮»
+q0 : ┤ h ├──●────●────●──┤ y ├┤ r1(3.142) ├──────╳─────╳──╳─────╳──●─┤>      ├»
+     ├───┤╭─┴─╮  │  ╭─┴─╮╰─┬─╯╰──┬─────┬──╯      │     │  │     │  │ │       │»
+q1 : ┤ h ├┤ x ├──●──┤ y ├──●─────┤ tdg ├─────────┼──╳──╳──┼──╳──╳──╳─┤●      ├»
+     ├───┤╰───╯╭─┴─╮╰─┬─╯  │     ╰┬───┬╯   ╭───╮ │  │     │  │  │  │ │  swap │»
+q2 : ┤ h ├─────┤ y ├──●────●──────┤ z ├────┤ s ├─╳──╳─────╳──╳──┼──╳─│       │»
+     ├───┤     ╰───╯              ╰───╯    ╰───╯                │  │ │       │»
+q3 : ┤ h ├──────────────────────────────────────────────────────●──●─┤>      ├»
+     ╰───╯                                                           ╰───────╯»
+
+################################################################################
+
+╭───────╮╭───────────╮    ╭─────╮   ╭────────────╮
+┤>      ├┤ rx(2.718) ├────┤ sdg ├───┤ rx(-2.718) ├
+│       │├───────────┤    ╰──┬──╯   ├────────────┤
+┤●      ├┤ ry(3.142) ├───────●──────┤ ry(-3.142) ├
+│  swap │├───────────┴╮╭───────────╮╰────────────╯
+┤●      ├┤ rz(-3.142) ├┤ rz(3.142) ├──────────────
+│       │╰────────────╯╰───────────╯              
+┤>      ├─────────────────────────────────────────
+╰───────╯                                         
+'''
+
+    # FIXME add this later
+    # assert circuit == expected_str
```

Comment:
Looks like there is an assert missing.

Reply:
Turns out we had a bug in lowering S<adj> and T<adj> to QIR, and that's why the I couldn't get the strings to equal each other.

Reply:
Good find!